fuzzing/ut_process/main.c

```diff
@@ -0,0 +1,41 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <err.h>
+
+#include "ut_process.h"
+
+#define BUFF_SIZE 1000
+
+
+int main(void)
+{
+    char input_buf[BUFF_SIZE];
+    char output_buf[BUFF_SIZE];
+
+    ssize_t input_len = read(STDIN_FILENO, input_buf, BUFF_SIZE);
+
+    if (input_len == -1) {
+        exit(EXIT_SUCCESS);
+        return 0;
+    }
+    if (input_len < BUFF_SIZE) {
+        input_buf[input_len] = '\0';
+    }
+    else {
+        input_buf[BUFF_SIZE - 1] = '\0';
+    }
+
+    /* typically looks like this: "http://www.example.com/foo{?query,number}" */
+    ut_process_var_t vars[] = {
+        { .name = input_buf, .value = &input_buf[11], },
+        { .name = &input_buf[22], .value = &input_buf[33], },
+    };
+    size_t res_buf_len = BUFF_SIZE;
+
+    ut_process_str_expand(&input_buf[44], vars, 2, output_buf, &res_buf_len);
```

Comment:
> Isn't it somewhat annoying [..]

Yes.

> Wouldn't it make more sense to supply vars[] separately?

Yes. 

> Not sure though if there is a good way to have AFL generate/supply multiple inputs.

Me neither. I looked a bit through the documentation & best practices without success. 
One approach could be using defining grammar as described here: [lcamtuf.blogspot.com/2015/01/afl-fuzz-making-up-grammar-with.html](https://lcamtuf.blogspot.com/2015/01/afl-fuzz-making-up-grammar-with.html). At the moment I'm not willig / able to go down that road :/ 

> How is null-termination of the value member enforced?

It's not. And from my perspective its a feature, otherwise AFL could not test non-null terminated input. 

(Also @nmeum in reply to that, GitHub external, translated from German)
> the fuzzer will generate a lot of inputs where there is no null terminator for the variable names and then you have a lot of inputs that are rejected early and where no meaningful exploration of this `ut_expand` function takes place (at least this is my assumption, which would also be seen in the coverage typically)

Reply:
> It's not. And from my perspective its a feature, otherwise AFL could not test non-null terminated input.

In this case null-termination of the `.value` strings is part of the `ut_process_str_expand` API contract, isn't it? Also: since the entire `input_buf` is null-terminated `.value` will also always be null-terminated it is just that it sometimes uses the entirety of the `input_buf` for `.value` then.